vertx-web/src/main/java/io/vertx/ext/web/handler/impl/AuthHandlerImpl.java

```diff
@@ -67,25 +86,168 @@ protected void authorise(User user, RoutingContext context) {
           if (res.result()) {
             if (count.incrementAndGet() == requiredcount) {
               // Has all required authorities
-              context.next();
+              handler.handle(Future.succeededFuture());
             }
           } else {
             if (sentFailure.compareAndSet(false, true)) {
-              context.fail(403);
+              handler.handle(Future.failedFuture(FORBIDDEN));
             }
           }
         } else {
-          context.fail(res.cause());
+          handler.handle(Future.failedFuture(res.cause()));
         }
       };
-      for (String authority: authorities) {
-        user.isAuthorised(authority, authHandler);
+      for (String authority : authorities) {
+        if (!sentFailure.get()) {
+          user.isAuthorised(authority, authHandler);
+        }
       }
     } else {
       // No auth required
-      context.next();
+      handler.handle(Future.succeededFuture());
     }
   }
 
+  protected String authenticateHeader(RoutingContext context) {
+    return null;
+  }
+
+  @Override
+  public void handle(RoutingContext ctx) {
+
+    if (handlePreflight(ctx)) {
```

Comment:
Not really, this was already present on JWT and OAuth2 but according to https://www.w3.org/TR/cors/#cross-origin-request-with-preflight-0 it should be for all authorization requests

I've just moved to the top level class since it is should be common